applications/GeoMechanicsApplication/custom_strategies/builder_and_solvers/residualbased_block_builder_and_solver_linear_elastic_dynamic.h

```diff
@@ -0,0 +1,703 @@
+            << "After the solution of the system"
+            << "\nSystem Matrix = " << rA << "\nUnknowns vector = " << rDx
+            << "\nRHS vector = " << rb << std::endl;
+    }
+
+    /**
+     * @brief This is a call to the linear system solver (taking into account some physical particularities of the problem)
+     * @param rA The LHS matrix
+     * @param rDx The Unknowns vector
+     * @param rb The RHS vector
+     * @param rModelPart The model part of the problem to solve
+     */
+    void InternalSystemSolveWithPhysics(TSystemMatrixType& rA, TSystemVectorType& rDx, TSystemVectorType& rb, ModelPart& rModelPart)
+    {
+        double norm_b = 0.00;
+        if (TSparseSpace::Size(rb) != 0) norm_b = TSparseSpace::TwoNorm(rb);
+
+        if (norm_b != 0.00) {
+            // if the system is already factorized, perform solution step. In case the solver does not support this, use Solve
+            if (mUsePerformSolutionStep) {
+                BaseType::mpLinearSystemSolver->PerformSolutionStep(rA, rDx, rb);
+            } else {
+                BaseType::mpLinearSystemSolver->Solve(rA, rDx, rb);
+            }
+        } else {
+            KRATOS_WARNING_IF("ResidualBasedBlockBuilderAndSolverLinearElasticDynamic",
+                              BaseType::mOptions.IsNot(SILENT_WARNINGS))
+                << "ATTENTION! setting the RHS to zero!" << std::endl;
+        }
+
+        // Prints information about the current time
+        KRATOS_INFO_IF("ResidualBasedBlockBuilderAndSolverLinearElasticDynamic", BaseType::GetEchoLevel() > 1)
+            << *(BaseType::mpLinearSystemSolver) << std::endl;
+    }
+
+    /**
+     * @brief Function to perform the build of the RHS.
+     * @param pScheme The integration scheme considered
+     * @param rModelPart The model part of the problem to solve
+     */
+    void BuildRHS(typename TSchemeType::Pointer pScheme, ModelPart& rModelPart, TSystemVectorType& rb) override
+    {
+        KRATOS_TRY
+
+        Timer::Start("BuildRHS");
+
+        this->BuildRHSNoDirichlet(pScheme, rModelPart, rb);
+
+        // add dirichlet conditions to RHS
+        this->ApplyDirichletConditionsRhs(rb);
+
+        Timer::Stop("BuildRHS");
+
+        KRATOS_CATCH("")
+    }
+
+    void CalculateReactions(typename TSchemeType::Pointer pScheme,
+                            ModelPart&                    rModelPart,
+                            TSystemMatrixType&            A,
+                            TSystemVectorType&            Dx,
+                            TSystemVectorType&            b) override
+    {
+        TSparseSpace::SetToZero(b);
+
+        // refresh RHS to have the correct reactions
+        this->BuildRHSNoDirichlet(pScheme, rModelPart, b);
+
+        // NOTE: dofs are assumed to be numbered consecutively in the BlockBuilderAndSolver
+        block_for_each(BaseType::mDofSet, [b](Dof<double>& rDof) {
```

Comment:
Let's pass this vector by ref to avoid a copy
```suggestion
        block_for_each(BaseType::mDofSet, [&b](Dof<double>& rDof) {
```